core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -424,79 +466,47 @@ protected boolean checkBlocks() {
                 }
                 return false;
             }
+            // request does not have a shardId yet, we need to pass the concrete index to resolve shardId
+            resolveRequest(clusterState.metaData(), concreteIndex, request);
+            assert request.shardId() != null : "request shardId must be set in resolveRequest";
             return true;
         }
 
-        protected ShardRouting resolvePrimary(ShardIterator shardIt) {
-            // no shardIt, might be in the case between index gateway recovery and shardIt initialization
-            ShardRouting shard;
-            while ((shard = shardIt.nextOrNull()) != null) {
-                // we only deal with primary shardIt here...
-                if (shard.primary()) {
-                    return shard;
-                }
-            }
-            return null;
-        }
+        private void performAction(final String nodeId, final String action, final boolean isPrimaryAction) {
+            DiscoveryNode node = observer.observedState().nodes().get(nodeId);
+            transportService.sendRequest(node, action, request, transportOptions, new BaseTransportResponseHandler<Response>() {
```

Comment:
I like this a lot! I restructured the doRun method in https://github.com/areek/elasticsearch/commit/1159d3ead370dcc94835b5b5146f7344eef65a0a to clarify that we use one copy of the state throughout the reroute phase. This change also nuked the  `checkBlocksAndResolveRequest` with side-effect that was pointed out in another comment.